ext-src/liveShare.ts

```diff
@@ -0,0 +1,136 @@
+    // further to initialize.
+    return;
+  }
+
+  await sessionStarted(liveShare);
+  setupServices(liveShare, windowManager);
+}
+
+async function setupServices(liveShare: vsls.LiveShare, windowManager: BrowserViewWindowManager) {
+  let service: vsls.SharedService | vsls.SharedServiceProxy;
+
+  if (liveShare.session.role === vsls.Role.Host) {
+    service = (await liveShare.shareService(SERVICE_NAME))!;
+
+    service.onRequest(REQUEST_GET_WINDOWS, () => {
+      const windows = Array.from(windowManager.openWindows);
+      return windows.map((window) => {
+        const viewport = window.browserPage!.page.viewport();
+
+        return {
+          startUrl: window.config.startUrl, // TODO: This needs to be the current URL
+          viewport
+        };
+      });
+    });
+  } else if (liveShare.session.role === vsls.Role.Guest) {
+    service = await liveShare.getSharedService(SERVICE_NAME);
+
+    if (!service) {
+      // The host doesn't have this extension installed
+      return;
+    }
+
+    const windows = await service.request(REQUEST_GET_WINDOWS, []);
+    if (windows && windows.length > 0) {
+      windows.forEach(({ startUrl, viewport }: any) => {
+        windowManager.create(startUrl, SHARED_PANEL_TITLE);
+      });
+    }
+
+    service.onNotify(NOTIFICATION_WINDOW_CREATED, async (args: any) => {
+      await windowManager.create(args.startUrl, SHARED_PANEL_TITLE);
+    });
+  }
+
+  handleRemoteInteractions(service, liveShare, windowManager);
+  handleLocalWindowCreation(service, liveShare, windowManager);
+}
+
+function sessionStarted(liveShare: vsls.LiveShare) {
+  return new Promise((resolve) => {
+    if (liveShare.session.id) {
+      resolve();
+    } else {
+      liveShare.onDidChangeSession((e) => {
+        if (e.session.id) {
+          resolve();
+        }
+      });
+    }
+  });
+}
+
+function handleRemoteInteractions(
+  service: vsls.SharedService | vsls.SharedServiceProxy,
+  liveShare: vsls.LiveShare,
+  windowManager: BrowserViewWindowManager
+) {
+  service.onNotify(NOTIFICATION_WINDOW_INTERACTION, async (args: any) => {
+    const { peerNumber, url, data } = args;
+
+    if (peerNumber === liveShare.session.peerNumber) {
+      // This is a re-broadcasted event from
+      // the same user.
+      return;
+    }
+
+    windowManager.openWindows.forEach((b: BrowserViewWindow) => {
+      if (b.config.startUrl == url) {
+        b.browserPage!.send(data.type, data.params);
+      }
+    });
+
+    if (liveShare.session.role === vsls.Role.Host) {
+      service.notify(NOTIFICATION_WINDOW_INTERACTION, args);
+    }
+  });
+}
+
+function handleLocalWindowCreation(
+  service: vsls.SharedService | vsls.SharedServiceProxy,
+  liveShare: vsls.LiveShare,
+  windowManager: BrowserViewWindowManager
+) {
+  windowManager.on(NOTIFICATION_WINDOW_CREATED, async (window: BrowserViewWindow) => {
+    if (liveShare.session.role === vsls.Role.Host) {
+      const viewport = await window.browserPage!.page.viewport();
```

Comment:
@auchenberg Is this the correct way to determine the width/height of the rendered preview window? In order to maintain the same dimensions on the guest-side, we need to grab this whenever a shared preview is created.

Reply:
Yes, we would need to extract that from the `viewportMetadata` which is kept in the front-end per https://github.com/auchenberg/vscode-browser-preview/blob/master/src/App.tsx#L10

One way could be to have an API to "getState" which then is transferred to the VS Code extension and shared for LiveShare. The same goes for the tab title, which is controlled by the front-end.